bin/evm2near/src/wasm_translate.rs

```diff
@@ -0,0 +1,714 @@
+                    for import in import_section {
+                        let import = import?;
+                        let typ = Translator::type_ref(import.ty)?;
+                        builder.imports.push(Import {
+                            module: import.module.to_string(),
+                            field: import.name.to_string(),
+                            ty: typ,
+                        });
+                    }
+                }
+                Payload::FunctionSection(function_section) => {
+                    for function in function_section {
+                        builder.functions.push(function?);
+                    }
+                }
+                Payload::TableSection(table_section) => {
+                    for table in table_section {
+                        let table = table?;
+                        let table_type = Translator::table_type(&table.ty)?; // todo TableInit is not used!
+                        builder.tables.push(table_type);
+                    }
+                }
+                Payload::MemorySection(memory_section) => {
+                    for mem in memory_section {
+                        let mem = mem?;
+                        let memory_type = Translator::memory_type(&mem)?;
+                        builder.memories.push(memory_type);
+                    }
+                }
+                Payload::TagSection(tag_section) => {
+                    for tag in tag_section {
+                        let tag = tag?;
+                        let _tag_type = Translator::tag_type(&tag)?;
+                        panic!("unspecified section (spec doesnt have that one! WTF)");
+                    }
+                }
+                Payload::GlobalSection(global_section) => {
+                    for glob in global_section {
+                        let (typ, init_instr) = Translator::global(glob?)?;
+                        builder.globals.push(Glob { typ, init_instr });
+                    }
+                }
+                Payload::ExportSection(export_section) => {
+                    for export in export_section {
+                        let export = export?;
+                        builder.exports.push(Export {
+                            name: export.name.to_string(),
+                            kind: Translator::ext_kind(export.kind),
+                            index: export.index,
+                        });
+                    }
+                }
+                Payload::StartSection {
+                    func: function_index,
+                    range: _range,
+                } => {
+                    let start_sect = wasm_encoder::StartSection { function_index };
+                    builder.start_sect = Some(start_sect);
+                }
+                Payload::ElementSection(element_section) => {
+                    for element in element_section {
+                        let element_segment = self.element(element?)?;
```

Comment:
here im getting error:
```
error[E0499]: cannot borrow `*self` as mutable more than once at a time
   --> bin/evm2near/src/wasm_translate.rs:325:47
    |
239 |     pub fn parse<'a>(&'a mut self, wasm: &'a Vec<u8>) -> Result<ModuleBuilder<'a>> {
    |                  -- lifetime `'a` defined here
...
325 |                         let element_segment = self.element(element?)?;
    |                                               ^^^^^^^^^^^^^^^^^^^^^^ `*self` was mutably borrowed here in the previous iteration of the loop
...
382 |         Ok(builder)
    |         ----------- returning this value requires that `*self` is borrowed for `'a`
```

I dont know how to fix it, but there is some [comments](https://users.rust-lang.org/t/mutably-borrowed-in-previous-iteration-despite-no-chance-of-conflicting-borrow/83274/2) that this problem will be fixed with new borrow checker

also, Im not sure if that approach to owning temporarily-created structs in translator object itself is what we need there

Reply:
I don't think moving the data into the translator object state is the right thing to do because then you need to keep the translator in scope for the entire duration you are working with the `ModuleBuilder`. I think it is better if the `ModuleBuilder` owns the data itself. Unfortunately this requires a bit of code duplication for some types in the `wasm_encoder` library, but I think it's ok for now because it is only a few types. We can remove this duplication in the future by changing the upstream types to use `Cow` (making an open source contribution to their library).



You can see my approach to getting rid of the `Box::leak` calls [here](https://github.com/MCJOHN974/evm2near/commit/9ddca5ca4ae82556a285a36e2ff6c36282d61e41).

I do like that in your changes many of the translator functions dropped the use of `self` where it was not needed. You can see in my approach that I introduced a second lifetime parameter instead to decouple from the translator, but I like your changes better in this respect.